File: go/text_bitmap.go
package main

// NewTextBitmap creates a bitmap that renders the given text string using the font.
func NewTextBitmap(font *Font, text string, color int, spacing int) *SimpleBitmap {
	cw := font.W
	ch := font.H
	numChars := len(text)
	if numChars == 0 {
		return NewSimpleBitmap(0, ch, nil)
	}
	fullWidth := (cw * numChars) + ((numChars - 1) * spacing)
	fullHeight := ch
	pixels := make([]int, fullWidth*fullHeight)

	for i := 0; i < numChars; i++ {
		charCode := int(text[i])
		cox := (cw + spacing) * i
		charPixels := font.PixelsForChar(charCode)
		for y := 0; y < ch; y++ {
			for x := 0; x < cw; x++ {
				if charPixels[y*cw+x] != 0 {
					pixels[fullWidth*y+cox+x] = color
				}
			}
		}
	}
	return NewSimpleBitmap(fullWidth, fullHeight, pixels)
}

File: php/pixelpong/bitmap/TextBitmap.php
<?php

namespace stigsb\pixelpong\bitmap;

class TextBitmap extends SimpleBitmap
{
    /**
     * @param Font $font
     * @param string $text
     * @param int $color
     * @param int $spacing
     * @throws FontException
     */
    public function __construct(Font $font, $text, $color, $spacing=1)
    {
        $cw = $font->getWidth();
        $ch = $font->getHeight();
        // no support for UTF-8 or newlines!
        $num_chars = strlen($text);
        $full_width = ($cw * $num_chars) + (($num_chars - 1) * $spacing);
        $full_height = $ch;
        $pixels = array_fill(0, $full_width * $full_height, 0);
        for ($i = 0; $i < $num_chars; ++$i) {
            $char = ord($text[$i]);
            $cox = ($cw + $spacing) * $i;
            $char_pixels = $font->getPixelsForCharacter($char);
            for ($y = 0; $y < $ch; ++$y) {
                for ($x = 0; $x < $cw; ++$x) {
                    if ($char_pixels[($y * $cw) + $x]) {
                        $pixels[($full_width * $y) + $cox + $x] = $color;
                    }
                }
            }
        }
        parent::__construct($full_width, $full_height, $pixels);
    }
}
